server/src/internalClusterTest/java/org/elasticsearch/action/support/master/TransportMasterNodeActionIT.java

```diff
@@ -107,27 +115,98 @@ public void onFailure(Exception e) {
+            safeAwait(stateApplierBarrier);
+            // Wait for the signal to unblock
+            safeAwait(stateApplierBarrier);
+            return null;
+        }, ActionListener.noop());
+        cleanupTasks.add(stateApplierBarrier::reset);
+
+        // Wait until state application is blocked
+        safeAwait(stateApplierBarrier);
+        return stateApplierBarrier;
+    }
+
+    /**
+     * Configure a latch that will be released when the existing master knows of the new master's election
+     *
+     * @param newMaster The name of the newMaster node
+     * @param cleanupTasks The list of cleanup tasks
+     * @return A latch that will be released when the old master acknowledges the new master's election
+     */
+    private CountDownLatch configureElectionLatch(String newMaster, List<Runnable> cleanupTasks) {
+        final String originalMasterName = internalCluster().getMasterName();
+        logger.info("Original master was {}, new master will be {}", originalMasterName, newMaster);
+        final var previousMasterKnowsNewMasterIsElectedLatch = new CountDownLatch(1);
+        ClusterStateApplier newMasterMonitor = event -> {
+            DiscoveryNode masterNode = event.state().nodes().getMasterNode();
+            if (masterNode != null && masterNode.getName().equals(newMaster)) {
+                previousMasterKnowsNewMasterIsElectedLatch.countDown();
+            }
+        };
+        ClusterService originalMasterClusterService = internalCluster().getInstance(ClusterService.class, originalMasterName);
+        originalMasterClusterService.addStateApplier(newMasterMonitor);
+        cleanupTasks.add(() -> originalMasterClusterService.removeApplier(newMasterMonitor));
+        return previousMasterKnowsNewMasterIsElectedLatch;
+    }
+
+    /**
+     * Add some master-only nodes and block until they've joined the cluster
+     *
+     * Ensure that we've got 5 voting nodes in the cluster, this means even if the original
+     * master accepts its own failed state update before standing down, we can still
+     * establish a quorum without its (or our own) join.
+     */
+    private static void createClusterOfSufficientSize() {
+        final var enoughVotingMastersLatch = new CountDownLatch(1);
+        ClusterStateApplier clusterFormationMonitor = event -> {
+            if (5 <= event.state().coordinationMetadata().getLastCommittedConfiguration().getNodeIds().size()) {
+                enoughVotingMastersLatch.countDown();
+            }
+        };
+        ClusterService newMasterClusterService = internalCluster().getAnyMasterNodeInstance(ClusterService.class);
+        try {
+            newMasterClusterService.addStateApplier(clusterFormationMonitor);
+            internalCluster().startMasterOnlyNodes(3);
+            safeAwait(enoughVotingMastersLatch);
+        } finally {
+            newMasterClusterService.removeApplier(clusterFormationMonitor);
+        }
+    }
```

Comment:
This feels a bit hacky, but I struggled to see how we could
1. Block the applier on newMaster to prevent it clearing the current master (in `becomeCandidate`)
2. Allow newMaster to send a joinRequest to itself via the cluster applier (in `JoinHelper`)
3. Block the applier on newMaster to prevent it applying the cluster state after it wins an election

I think (although with all the asynchrony, I'm not sure) that the above things happen in the above order, so I don't know how we could block 1 and 3 whilst allowing 2 to go through - and if we could, would the extra machinery just make the test harder to read and more brittle?

That means we can't rely on our own join, and because the old master sometimes(?) accepts its failed update, we can't rely on its join, so we can't reliably get a quorum in a 3 node cluster. If we bump the cluster size up to 5 nodes we seem to succeed every time.